sql/core/src/main/scala/org/apache/spark/sql/sources/commands.scala

```diff
@@ -41,3 +55,359 @@ private[sql] case class InsertIntoDataSource(
+        DataFrame(sqlContext, query).queryExecution.toRdd,
+        relation.schema,
+        needsConversion = false)
+
+      if (partitionColumns.isEmpty) {
+        insert(new DefaultWriterContainer(relation, job), df)
+      } else {
+        val writerContainer = new DynamicPartitionWriterContainer(
+          relation, job, partitionColumns, "__HIVE_DEFAULT_PARTITION__")
+        insertWithDynamicPartitions(writerContainer, df, partitionColumns)
+      }
+    }
+
+    Seq.empty[Row]
+  }
+
+  private def insert(writerContainer: BaseWriterContainer, df: DataFrame): Unit = {
+    try {
+      writerContainer.driverSideSetup()
+      df.sqlContext.sparkContext.runJob(df.queryExecution.executedPlan.execute(), writeRows _)
+      writerContainer.commitJob()
+      relation.refresh()
+    } catch { case cause: Throwable =>
+      writerContainer.abortJob()
+      throw new SparkException("Job aborted.", cause)
+    }
+
+    def writeRows(taskContext: TaskContext, iterator: Iterator[Row]): Unit = {
+      writerContainer.executorSideSetup(taskContext)
+
+      try {
+        while (iterator.hasNext) {
+          val row = iterator.next()
+          writerContainer.outputWriterForRow(row).write(row)
+        }
+        writerContainer.commitTask()
+      } catch { case cause: Throwable =>
+        writerContainer.abortTask()
+        throw new SparkException("Task failed while writing rows.", cause)
+      }
+    }
+  }
+
+  private def insertWithDynamicPartitions(
+      writerContainer: BaseWriterContainer,
+      df: DataFrame,
+      partitionColumns: Array[String]): Unit = {
+
+    require(
+      df.schema == relation.schema,
+      s"""DataFrame must have the same schema as the relation to which is inserted.
+         |DataFrame schema: ${df.schema}
+         |Relation schema: ${relation.schema}
+       """.stripMargin)
+
+    val sqlContext = df.sqlContext
+
+    val (partitionRDD, dataRDD) = {
+      val fieldNames = relation.schema.fieldNames
+      val dataCols = fieldNames.filterNot(partitionColumns.contains)
+      val df = sqlContext.createDataFrame(
+        DataFrame(sqlContext, query).queryExecution.toRdd,
+        relation.schema,
+        needsConversion = false)
+
+      val partitionColumnsInSpec = relation.partitionSpec.partitionColumns.map(_.name)
+      require(
+        partitionColumnsInSpec.sameElements(partitionColumns),
+        s"""Partition columns mismatch.
+           |Expected: ${partitionColumnsInSpec.mkString(", ")}
+           |Actual: ${partitionColumns.mkString(", ")}
+         """.stripMargin)
+
+      val partitionDF = df.select(partitionColumns.head, partitionColumns.tail: _*)
+      val dataDF = df.select(dataCols.head, dataCols.tail: _*)
+
+      partitionDF.queryExecution.executedPlan.execute() ->
+        dataDF.queryExecution.executedPlan.execute()
+    }
+
+    try {
+      writerContainer.driverSideSetup()
+      sqlContext.sparkContext.runJob(partitionRDD.zip(dataRDD), writeRows _)
```

Comment:
Maybe it is better to use a projection to extract values of those partition columns?